Unit testing test output

sonicyeti opened the pull request:
This PR covers the creation of a list group & supporting list group item specifically for the the purpose of our unit testing right rail which is responsible for test result display. Styling provisioning for an expand & collapse chevron has been implemented for next iteration.

Context, see item 10: https://github.com/Kong/insomnia/issues/2464

**Before:**
![Screen Shot 2020-08-20 at 8 18 50 AM](https://user-images.githubusercontent.com/52717970/90769710-acdc3e00-e2be-11ea-8e3d-a7a0b4a25ef4.png)

**After:**
![Screen Shot 2020-08-20 at 8 23 44 AM](https://user-images.githubusercontent.com/52717970/90769718-b1085b80-e2be-11ea-8223-51b8edc8fdf7.png)

**Supporting Story:**
![Screen Shot 2020-08-20 at 8 26 13 AM](https://user-images.githubusercontent.com/52717970/90769806-d4cba180-e2be-11ea-821b-62fa5a4f09a8.png)

sonicyeti committed:
Adding list group component with stubbed badge & list item

sonicyeti committed:
Cleaning up stories, error padding tweak

sonicyeti committed:
Results list item padding fix

develohpanda commented on packages/insomnia-components/components/list-group-item.js:
```diff
@@ -0,0 +1,115 @@
+  }
+`;
+
+// TODO: Create a collection of visual badges within our component system.
```
We have tags (like packages/insomnia-app/app/ui/components/tags/method-tag.js). Is that similar?

sonicyeti replied:
Similar in theory but the design is very far off and more appropriate for accent type usages (currently). My hope is that we can craft a nice array of badges similar to other frameworks.

develohpanda commented on packages/insomnia-components/components/list-group-item.js:
```diff
@@ -0,0 +1,115 @@
+  }
+`;
+
+class ListGroupItem extends React.PureComponent<Props> {
```
This is named as a really generic component, but it's actually _very_ specific to test results. It almost doesn't need to be within `insomnia-components`, but at the very least I would name it more specifically and/or separate what is needed for a `ListGroupItem` (eg expand/collapse) vs what is needed for a `TestResultItem`

develohpanda commented on packages/insomnia-components/components/list-group-item.js:
```diff
@@ -0,0 +1,115 @@
+    }
+  }
+
+  .itemToggler {
```
Could we please remove any direct classes from a styled-component in favor of a standalone styled-component? See `notice-table` for an example.

You can also [extend styles](https://styled-components.com/docs/basics#extending-styles) if need be

develohpanda commented on packages/insomnia-components/components/list-group-item.js:
```diff
@@ -0,0 +1,115 @@
+class ListGroupItem extends React.PureComponent<Props> {
+  render() {
+    const { errorMsg, result, name, time } = this.props;
+    return (
```
Great candidate for a functional component 😄

develohpanda commented on packages/insomnia-components/components/list-group.js:
```diff
@@ -0,0 +1,23 @@
+  padding: 0px var(--padding-xs) 0px 0px;
+`;
+
+class ListGroup extends React.PureComponent<Props> {
```
Great candidate for a functional component 😄

develohpanda commented on packages/insomnia-app/app/ui/components/wrapper-unit-test.js:
```diff
@@ -327,19 +336,25 @@ class WrapperUnitTest extends React.PureComponent<Props, State> {
-                      <br />
-                      <code className="text-danger">{t.err.message}</code>
-                    </>
+                    <ListGroupItem
+                      key={i}
+                      name={t.title}
+                      time={t.duration}
+                      result="fail"
+                      errorMsg={t.err.message}
+                    />
                   )}
-                </li>
+
+                  {!t.err.message && (
+                    <ListGroupItem key={i} name={t.title} time={t.duration} result="pass" />
+                  )}
```
This logic can (and should, IMO), be inside a `TestResultListItem` component.

develohpanda reviewed (commented):
Couple of notes :)

DMarby commented on packages/insomnia-components/components/list-group.stories.js:
```diff
@@ -0,0 +1,26 @@
+    <ListGroup>
+      <ListGroupItem
+        name="My test name one"
+        time="602ms"
```
Nitpick, but imo it looks kind of strange to have the unit directly after the number, I think adding a space to make it `602 ms` would make more sense, same for all the other items.

DMarby commented on packages/insomnia-components/components/list-group-item.js:
```diff
@@ -0,0 +1,115 @@
+        <div>
+          <div className="itemToggler">
+            {/* 
+            TODO: Add show/hide toggle support in the future.
```
Imo it's nice to avoid leaving comments/commented code like this as much as possible, if we want something to be done in the future, we should open an issue for it to have it properly tracked instead.

Not sure what the common approach is in this codebase however, but that's my two cents.

DMarby commented on packages/insomnia-components/components/list-group-item.js:
```diff
@@ -0,0 +1,115 @@
+          <p>{name}</p>
+          <div className="timestamp">
+            <SvgIcon icon={IconEnum.clock}></SvgIcon>
+            <span>{time}</span>
```
Since this is just outputting the raw input, currently, we already have two different results here:
- The storybook has strings in the format of `100ms`
- The actual UI has integers, resulting in just `100`

I think it would be better to just always pass in milliseconds or a duration type or such to the component, and have the presentation formatting be done by the component itself, this way it's always consistent.
With presentation formatting I mean either doing something like `%d ms` which is equivalent to what we had before this change, or by using a library that rounds up to seconds/minutes/etc.

develohpanda replied:
To add to that, I suggest naming the prop as `timeMilliseconds` or `timeMs`. Almost everything else in JS operates with ms, so lets stick with that convention, both with naming and with logic.

sonicyeti committed:
Review feedback and component provisioning

Shift to functional components
Abstract both list and list item away for re-use
Roll unit test result specific styling into UT source
Roll supporting UI over including badges > extend SC for flavors
Shift timestamp to render the value, JSX handling formatting

sonicyeti committed:
Renaming UT specific badges

sonicyeti committed:
List group and unit test wrapper component updates

- Pulled unit test results list item into SB
- All unique aspects of UT results are bundled in the IC which extend base item
- Cleansed UT wrapper of inline components
- Setting up list group and list group item base for middle area

sonicyeti committed:
Abstrating badges, timestamp and updating unit test wrapper

sonicyeti committed:
Merge branch 'develop' into unit-testing-test-output

develohpanda commented on packages/insomnia-components/components/list-group/list-group.stories.js:
```diff
@@ -0,0 +1,58 @@
+      ))}
+    </ListGroup>
+  </div>
+);
```
Epic!

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-result-badge.js:
```diff
@@ -0,0 +1,39 @@
+import styled from 'styled-components';
+
+type Props = {|
+  result?: String,
```
I suggest changing this prop to be more explicit

```suggestion
  failed?: boolean,
```

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-result-item.js:
```diff
@@ -0,0 +1,54 @@
+  return (
+    <StyledResultListItem>
+      <div>
+        <UnitTestResultBadge result={item.err.message} />
```
Then updating this usage:

```suggestion
        <UnitTestResultBadge failed={!!item.err.message} />
```

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-result-item.js:
```diff
@@ -0,0 +1,54 @@
+  );
+};
+
+export { UnitTestResultItem, UnitTestResultBadge, UnitTestResultTimeStamp, SvgIcon };
```
Do all of these still need to be exported? I would think we only need to export `UnitTestResultItem` now.

If all do need to be exported, careful about re-exporting `SvgIcon`

sonicyeti replied:
Thanks for the catch!

develohpanda commented on packages/insomnia-components/components/list-group/unit-test-result-timestamp.js:
```diff
@@ -0,0 +1,36 @@
+
+type Props = {|
+  timeMs: String,
+  children?: React.Node,
```
`children` prop can be removed now 😄

develohpanda commented on packages/insomnia-components/index.js:
```diff
@@ -7,13 +7,16 @@ import _DropdownDivider from './components/dropdown/dropdown-divider';
 import _RadioButtonGroup from './components/radio-button-group';
 import _Switch from './components/switch';
 import _ToggleSwitch from './components/toggle-switch';
+import * as unitTestResultItem from './components/list-group/unit-test-result-item';
```
After updating the exports from `unit-test-result-item`, we should be able to change this back to follow the same pattern as the other exports.

develohpanda reviewed (changes requested):
Almost there!! A couple of minor notes 😄 looking really good now.

sonicyeti committed:
Update packages/insomnia-components/components/list-group/unit-test-result-badge.js

Co-authored-by: Opender Singh <opender94@gmail.com>

sonicyeti committed:
Update packages/insomnia-components/components/list-group/unit-test-result-item.js

Co-authored-by: Opender Singh <opender94@gmail.com>

sonicyeti committed:
Dep cleanup, PR feedback
